Add support for the Cart & Checkout in WooCommerce's declare compatibility feature

tarhi-saad opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

For a more detailed description, please read the one in this ticket #36413!

In summary, we added the Cart & Checkout Blocks to the list of features so that other extension developers might declare (in)compatibility with it. For the time being, we are preventing the display of the Cart & Checkout Blocks feature's UI settings. We may choose otherwise in the future, but for the time being, we only need the Cart & Checkout Blocks to be included in the list of WooCommerce features.

Closes #36413.

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

Checkout to this branch, then:

1. Create a new plugin or clone an existing one, install it, and activate it on your test website
2. Declare its compatibility with the Cart & Checkout Blocks by adding the following snippet into the main plugin file:

```PHP

add_action( 'before_woocommerce_init', function() {
	if ( class_exists( \Automattic\WooCommerce\Utilities\FeaturesUtil::class ) ) {
		\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'cart_checkout_blocks', __FILE__, true );
	}
} );

```

3. Clone the [WooCommerce Blocks](https://github.com/woocommerce/woocommerce-blocks) plugin, install it, and activate it in your test website.
4. Check if we can get the list of compatible plugins with the Cart & Checkout Blocks with the WooCommerce Blocks plugin, and that the name of the plugin from step 1 is included. We can use the function `get_compatible_plugins_for_feature` from the `FeaturesController` API. For example, we can add the following snippet within the main plugin file:

```PHP

add_action(
	'woocommerce_init',
	function() {
		if ( class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) ) {
			var_dump(
				\Automattic\WooCommerce\Utilities\FeaturesUtil::get_compatible_plugins_for_feature( 'cart_checkout_blocks' )
			);
		}
	}
);

```

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

tarhi-saad committed:
Add the Cart & Checkout Blocks as a feature

tarhi-saad committed:
Prevent displaying UI setting of the C&C Blocks

Currently, we don't desire displaying the C&C Blocks feature's settings.
We may choose otherwise in the future, but for the time being adding
the C&C Block to the list of features does the work!

tarhi-saad committed:
Add a change file

tarhi-saad committed:
Fix PHPCS lint error: short array syntax

opr commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -535,7 +540,7 @@ function( $feature_id ) use ( $features ) {
 			}
 		);
-		$mature_feature_ids       = array_diff( $feature_ids, $experimental_feature_ids );
+		$mature_feature_ids       = array_diff( $feature_ids, $experimental_feature_ids, array( 'cart_checkout_blocks' ) );
```
Isn't `cart_checkout_blocks` a key of `$features` already, meaning `$feature_ids` would include it? Is there a reason you had to specifically add it here?

tarhi-saad replied:
Hi @opr! Thank you for the review! 🙌 

> Isn't cart_checkout_blocks a key of $features already, meaning $feature_ids would include it? Is there a reason you had to specifically add it here?

In this line, we are removing the `cart_checkout_blocks` to exclude it from the `mature_feature_ids` list. We don't want the add a `setting UI` for the Cart & Checkout feature. We will get the following result without this line:

<img width="1123" alt="image" src="https://user-images.githubusercontent.com/14235870/212893943-09d91df5-5741-4c85-9ad8-089e94c12558.png">

opr replied:
Great, thanks for explaining!

opr reviewed (commented):
Hey Saad, thanks for working on this. The functionality is working fine! Just one comment that needs to be resolved from my end.

jorgeatorres commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -535,7 +540,7 @@ function( $feature_id ) use ( $features ) {
 			}
 		);
-		$mature_feature_ids       = array_diff( $feature_ids, $experimental_feature_ids );
+		$mature_feature_ids       = array_diff( $feature_ids, $experimental_feature_ids, array( 'cart_checkout_blocks' ) );
```
While this works for what you intend it to, I feel we should have a more generic approach to adding features that don't need (or want) to expose a setting in the UI. For example, similar to the `is_experimental` flag we could have a `disable_ui` flag or something like that and then we would exclude those settings from `$feature_ids` below.

I'd be more than happy to review that change if you want to include it in this PR. If not, that's also fine, and we can add it later via a different PR.

Let me know what your thoughts are.

tarhi-saad replied:
This is an awesome suggestion! 💯 Thank you, @jorgeatorres! 🙌

jorgeatorres commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -105,6 +105,11 @@ public function __construct() {
 				'description'     => __( 'Enable the high performance order storage feature.', 'woocommerce' ),
 				'is_experimental' => true,
 			),
+			'cart_checkout_blocks' => array(
```
The arrows here are not correctly aligned resulting in a PHPCS error: https://github.com/woocommerce/woocommerce/actions/runs/3928725851/jobs/6716673300#step:6:20.

tarhi-saad replied:
Nice catch! 👍

It's weird! Running PHP lint locally shows me different results than the ones detected in CI.

jorgeatorres reviewed (changes requested):
Hi @tarhi-saad!

Overall, this looks good, though there's a tiny PHPCS issue that needs to be fixed. I also left some comments re: the approach to hiding the UI, but that's not necessarily a blocker. Please let me know what you think!

tarhi-saad committed:
Fix PHPCS lint error

tarhi-saad committed:
Add generic approach to adding features without UI

In our case for the Cart & Checkout feature, we don't want to show the
UI in the "Features" tab in Woo Settings. Creating a flag for this
purpose will make it easier for future features to unsubscribe from
showing a UI if required

tarhi-saad committed:
Merge branch 'woocommerce:trunk' into add/36413-support-for-cart-checkout-in-declare-compatibility

tarhi-saad committed:
Fix PHPCS errors

tarhi-saad committed:
Fix "undefined `disable_ui` index" error

tarhi-saad commented:
Hello, @jorgeatorres! 👋 Thank you for the review. And I apologize for addressing your feedback this late. Last week I was release manager, and I couldn't find the time for my ongoing tasks! Please could you give it another review?

jorgeatorres commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -545,7 +555,13 @@ function( $feature_id ) use ( $features ) {
+				return $features[ $feature_id ]['disable_ui'];
+			}
+		);
+		$mature_feature_ids       = array_diff( $feature_ids, $experimental_feature_ids, $disabled_ui_feature_ids );
```
We should not exclude features with disabled UI here as `$mature_feature_ids` as that wouldn't prevent experimental features with `disable_ui = true` from having their settings rendered.

My personal suggestion, would be to add a check on [this line](https://github.com/woocommerce/woocommerce/blob/615365d14e0e2b944f05213e366fd1a9849612db/plugins/woocommerce/src/Internal/Features/FeaturesController.php#L595) such as

```php
if ( isset( $features[ $id ]['disable_ui'] ) && $features[ $id ]['disable_ui'] ) {
	continue;
}
```

That way, we can skip displaying the setting for features with disabled UI whether it is a mature or experimental feature.

As a side effect, we can get rid of the `$disabled_ui_feature_ids` var, at least for the time being.

jorgeatorres reviewed (changes requested):
Hi @tarhi-saad!

Thanks for making the changes. I noticed a tiny issue with the way `disable_ui` filtering was implemented, and so I left a suggestion in the code. Other than that, this is looking great 💯.

Please let me know what you think.

tarhi-saad committed:
Apply disable UI filter to experimental features too

This change disable UI for both experimental and mature features

Previously we applied this filter to mature ones only

jorgeatorres reviewed (approved):
@tarhi-saad: This looks good to me. Thanks for making the changes! 💯

I see you requested a review from others, so I'm not sure if you want to wait until they review. If not, feel free to go ahead with the merging part (or ping me if you're not allowed to do that).

tarhi-saad commented:
Thank you, @jorgeatorres, for your review and the PR approval! 🙌 Your feedback was incredibly valuable. Thanks to your suggestions, the code is now looking better than ever! I'll go ahead and merge it!
